discovery_engine_core/providers/src/client.rs

```diff
@@ -12,37 +12,256 @@
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-//! Client do get new documents.
+//! Client to get new documents.
+
+use std::collections::HashMap;
+
+use displaydoc::Display as DisplayDoc;
+use thiserror::Error;
 
 use xayn_discovery_engine_core::document::Document;
 
-use crate::{filter::Filter, query::Query};
+use crate::{filter::Filter, newscatcher::Response as NewscatcherResponse};
 
-// TODO: replace with error form the http library + Filter error
-pub enum Error {}
+#[derive(Error, Debug, DisplayDoc)]
+pub enum Error {
+    /// Failed to execute the HTTP request: {0}
+    RequestExecution(#[source] reqwest::Error),
+    /// Server returned a non-successful status code: {0}
+    StatusCode(#[source] reqwest::Error),
+    /// Failed to decode the server's response: {0}
+    Decoding(#[source] reqwest::Error),
+}
 
 /// Client that can provide documents.
 pub struct Client {
     token: String,
     url: String,
 }
 
+/// The only accepted format is ISO 639-1 — 2 letter code, e.g. 'de' or 'en'
+pub(crate) type Language = String;
+/// The only acceptable format is ISO 3166-1 alpha-2, e.g. 'DE' or 'AU'
+pub(crate) type Country = String;
+
+/// Parameters determining which news to fetch
+pub struct NewsQuery {
+    language: Language,
+    country: Country,
+    filter: Filter,
+    /// How many articles to return (per page).
+    page_size: Option<usize>,
+}
+
+/// Parameters determining which headlines to fetch
+pub struct HeadlinesQuery {
+    language: Language,
+    country: Country,
+    /// How many articles to return (per page).
+    page_size: Option<usize>,
+}
+
 impl Client {
     /// Create a client.
     pub fn new(token: String, url: String) -> Self {
         Self { token, url }
     }
 
-    /// Get document from a provider,
-    ///
-    /// `filter` can be used to query only documents of interests,
-    /// `size` optionally limit the number of item we want.
-    pub async fn get(&self, filter: &Filter, size: Option<u8>) -> Result<Vec<Document>, Error> {
-        let _query = Query::new(self.token.clone(), filter, size);
+    /// Retrieve news from the remote API
+    pub async fn news(&self, params: &NewsQuery) -> Result<Vec<Document>, Error> {
+        let mut query: HashMap<String, String> = HashMap::new();
+        query.insert("sort_by".into(), "relevancy".into());
+        Self::build_news_query(&mut query, params);
+
+        let c = reqwest::Client::new();
+        let response = c
+            .get(format!("{}/v2/search", self.url))
+            .header("x-api-key", &self.token)
+            .query(&query)
+            .send()
+            .await
+            .map_err(Error::RequestExecution)?
+            .error_for_status()
+            .map_err(Error::StatusCode)?;
+
+        let news: NewscatcherResponse = response.json().await.map_err(Error::Decoding)?;
+        let result: Vec<Document> = news.articles.iter().map(|article| article.into()).collect();
```

Comment:
We cannot return a `Document` from here then. To create a `Document` we need the smbert. I think we can then return the `Vec<Article>`?
This could also make sense because then we can optimize a bit better on which document we need to run it.
We can a set of filter that does not depend on it that we can run and the compute the smbert only on the remaining.
@janpetschexain